### ISSUE TITLE
[build] Linux make install sandbox

### ISSUE THREAD

**Issue by RDR8** [raysan5/raylib#468]
I'm looking for the best general method to install and build against a personal working copy of raylib as an unprivileged user. The use case includes configuring, extending and generally working on libraylib directly in a student lab scenario. This initial case is in a computer science context with the learning goals being everything to do with C that isn't C. Tools like gcc, cpp, ld, make and gdb will all be exposed and explored before moving on to the after-party and games . I'd like to enable easy building and selection of various personal configurations of raylib in parallel. For example, one might want to have both OpenGL2.1 and OpenGLES versions installed side by side on a Raspberry Pi then select from them at build or run time. The current makefiles accommodate this but it is an improvised and unfinished solution. At this time, my primary concern is avoiding root appropriately. The check for root in the install target could by isolated to a SYSTEM mode with normal installation to the home or custom directory defaulting to a USER mode without root. I'd like to survey how others manage this scenario and if the implementation could be brought in line with best practices. Thanks

**Comment by a3f** [comment]
[`Alien::raylib`](https://github.com/athreef/Alien-raylib/blob/master/alienfile) does that. If you got `cpan` (which is part of the standard Perl installation on many distros), you can just write `cpan Alien::raylib` and it will:

* Check with `pkg-config --modversion raylib` if raylib is installed
* if not, clone the github repo, build it with CMake and install shared and static libraries it into the user's home directory
* Install an `Alien::raylib` module that wraps the system or user raylib

So, one could just

```
cpan Alien::libvas
echo "PKG_CONFIG_PATH=$(perl -MAlien::raylib -e 'print Alien::raylib->dist_dir')/lib/pkgconfig" >> ~/.bashrc
```

And you got yourself a locally installed raylib with which you can do:
```
cc game.c `pkg-config --cflags --libs raylib`
or cc game.c `pkg-config --cflags --libs --static raylib`
```

**Comment by OvermindDL1** [comment]
As raylib is a cmake project, just tell it to 'install' somewhere else, like if you want to install it to `$HOME/libs` then when you run cmake just add `-DCMAKE_INSTALL_PREFIX="$HOME/libs"`, or you can change the install location at `make` time just by adding `DESTDIR="$HOME/libs"` to the `make` command (assuming you are using make, other build systems have their own methods, or just use the cmake one that sets it for all of them regardless).  At which point just build and install as normal with any other cmake project.  :-)

**Comment by RDR8** [comment]
@a3f Maybe you could write a Perl HOWTO describing your process? How do you get the examples find raylib at runtime? I can build locally with the Makefiles but need sudo ldconfig $(RAYLIB_INSTALL_PATH) to enable system-wide runtime linking to raylib installed to a non-standard location, i.e. ~/lib. If I want to enable runtime linking without sudo, in examples/Makefile I append -Wl,-rpath,$(RAYLIB_RUNTIME_PATH) to CFLAGS, embedding the location directly in the executable. I haven't settled on a default switch  path. RAYLIB_RELEASE_PATH lets us work with the staged shared libs without formal installation. It's definition needs to be refined so that it is a full path.

**Comment by a3f** [comment]
@RDR8 That HowTo would be rather short, it just uses [CMake and pkg-config](https://github.com/athreef/Alien-raylib/blob/master/alienfile). I don't think trying to get feature parity between CMake and the Makefiles is a worthwhile effort.

**Comment by RDR8** [comment]
Thank you both for your input. My perspective on make vs cmake, is that make is the more fundamental tool for someone just learning this kind of software development, which is what I am and the "user" I'm catering to. I'm really interested in understanding how these basic tools work and in leaving some breadcrumbs for anyone who may follow. I guess it's kind of important to be correct in a public context, which is why I'm soliciting comments so loudly. CMake seems essential too and obviously works for a lot of people and it's on my need-to-know list. I just haven't gotten to it but I do appreciate having the option. For now, I need to learn about special variables in general.

Regarding RAYLIB_INSTALL_PATH and RAYLIB_RUNTIME_PATH, I guess they're being used like DESTDIR or BINDIR type variables. I like explicitly naming things but maybe I'm losing some features... Anyhow, in the current examples/Makefile, RAYLIB_RUNTIME_PATH gets embedded in our executable RPATH. This bypasses the need to "install" raylib with `sudo ldconfig`. I did a little research and this method seems to be preferred over LD_LIBRARY_PATH. It defaults to RAYLIB_INSTALL_PATH but offers a way to deploy applications, pointing to a bundled copy of libraylib.so. I don't know if anyone distributes games this way or just builds static distributables? I know it's not really the Unix way but it makes prototyping a heck of a lot easier. I don't want to mess with my system too much while pinching these libraries, you know?

To recap, I want to enable an edit-build-debug workflow that is suitable for beginners. It cannot require root permissions. It must reflect sensible, if not best, development practices leading to greater understanding of the GNU application development environment. It must not interfere with existing configuration of other platforms. I'll use a combination of configuration and documentation, mostly documentation, to get this done. This roughly corresponds to the first phase of raysan5's curriculum. The games are the bait and the bonus. Hopefully, in a year or two, with you guys' help, I'll be slinging C like it ain't no thang. Writing helps me understand things. Thanks for reading.

**Comment by OvermindDL1** [comment]
> To recap, I want to develop an edit-build-debug workflow that is suitable for beginners. It cannot require root permissions. It must reflect sensible, if not best, development practices leading to greater understanding of the GNU application development environment. It must not interfere with existing configuration of other platforms. I suggest a combination of configuration and documentation, mostly documentation, to achieve that. This roughly corresponds to the first phase of raysan5's curriculum. The games are the bait and the bonus. Hopefully, in a year or two, with you guys' help, I'll be slinging C like it ain't no thang. Writing helps me understand things. Thanks for reading.

Well the usual way to build an application in modern 'best pratices' is:

1 `git clone <blah>` Grab the project of course, but don't go in it's directory
2. `mkdir whereever; cd whereever` Make a directory to put all your build stuff in and go in to it
3. `cmake ../blah` Or whatever path to the project, it can be useful to pass options like:
  * `-DCMAKE_BUILD_TYPE=Release` to enable full release building, debug info is stripped, etc...
  * `-DCMAKE_INSTALL_PREFIX="$HOME/whatever"` to 'install' it to some place, otherwise it defaults to something like `/usr/local` or whatever is usually the default on the system being used.
  * `-G Ninja` You can choose other generators as well, default on linux is `"Unix Makefiles"`, but `Ninja` is a popular once considering how blazing *FAST* it is.  On mac this defaults to xcode I think, and on windows I think this defaults to Visual Studio nmake files unless run in cygwin or so where it defaults to `"Unix Makefiles"`, but you can of course generate any generator on any OS, so you can generate xcode or VStudio ones or whatever from linux if you want (though not that useful of course).  I tend to like ninja for it's absolute speed, but I'll use `make` for the remaining examples as it is the oldest and most documented.
  * There are others as well, this is like the normal `./configure ...` step in normal `Makefiles` (except it supports out of source builds, which you should always do).
  * You can make other paths and run `cmake <path-to-project>` with different options as well, so you can have a debug mode, a release mode, a release+debuginfo mode, maybe enable some options and libraries, disable others, etc... (this is why out of source builds are so useful, you cannot do a variety with in-source builds, in addition to corrupting the build tree of course).
4. Once configured then you can perform any of these steps:
  * `make` Build the project, this won't install or anything, you can of course pass the usual arguments like `DESTDIR` and so forth to override more if you want (if you run install that is), or things like `-j9` or so to parallize the builds or so.
  * `make test` Run the tests, if the project had tests of course.  In addition to compiling the tests this will also compile the main project like if you ran `make`.
  * `make install` Run the installer, it will install to where-ever specified, this also builds the project like if you ran `make` first.
  * There are others as well like package and so forth, in addition to potentially project-specific ones.
5. Instead of running whatever commands straight, you can also run it via cmake commands, so you could replace section 4 above with these commands instead and it would do the same thing (but not just for make but rather whatever generator was specified):
  * `cmake --build .` Build the project, you can specify a path other than `.` if you are not already in the configured directory too, with this way you can build many configurations at once too
  * `cmake --build . --target test` Run the tests
  * `cmake --build . --target install` Install as specified

In general with most well designed projects the CMake file will handle it's own dependencies, either picking it up from the system (CMake supports both cmake config files as well as pkg-config setups) or many projects will build dependencies internally (CMake can git clone and download all such as well).  The more poorly made projects will require you to manually download and build other parts (PR those!  That is just irritating...).  Things like [hunter](https://docs.hunter.sh/) (a CMake per-project addin) can handle the dependency management for the project, significantly reducing the dependency handling code to generally only a line or two (but many still do it manually for whatever reason).

CMake projects can also have toolchains, these are generally used when cross-compiling, so like compiling to android, rpi, compiling to windows from linux, etc... etc...  This is specified on the `cmake ...` step as well.

In general, most of the times you will just do this with well set up completed projects:

```zsh
$ git clone ... blah
$ mkdir blah-build-release && cd blah-build-release
$ cmake ../blah ...
$ cmake --build . --target install
$ # Repeat the last 3 steps for each different configuration that you want, can also do them all at the same time with simple setup too
```

**Comment by RDR8** [comment]
@OvermindDL1 Thank you very much for the thoughtful and detailed response. As much as I like to read, the experience of others is invaluably expeditious (^; It'll get a few re-reads. Thanks again.

No one's compaining about -rpath yet, so that's good. I think I'll try to feature it along with a hypothetical RAYLIB_INSTALL_MODE=USER

**Comment by OvermindDL1** [comment]
> @OvermindDL1 Thank you very much for the thoughtful and detailed response. As much as I like to read, the experience of others is invaluably expeditious (^; It'll get re-read thoroughly. Thanks again.

Heh, really though if you want to learn cmake, the [Official CMake manual](https://cmake.org/documentation/) and the [Hitchhiker's Guide to CMake](https://cgold.readthedocs.io/) (made by one of the big CMake people) are absolutely invaluable.  The equivalent CMake scripts compared to Makefiles are shorter and more readable (except in the trivial single-liner makefile cases, which are dreadfully rare), in addition you can [get it to work on any OS and build system](https://github.com/ruslo/polly) with relative ease (*especially* compared to Makefiles or visual studio) among many other things (like pure-cmake dependency systems like [Hunter](https://docs.hunter.sh/) or standalone needs-to-be-installed dependency systems like [Conan](https://conan.io/), both are distributed, out-of-source, etc... etc..., though the nice thing with a project that uses hunter is you need nothing but build-essentials and cmake to build them ^.^).  It is *well* worth learning.  :-)

**Comment by RDR8** [comment]
Taking a step with #480 .
